keep/api/core/db.py

```diff
@@ -430,6 +435,94 @@ def get_workflows_that_should_run():
         return workflows_to_run
 
 
+def update_workflow_by_id(
+    id,
+    name,
+    tenant_id,
+    description,
+    interval,
+    workflow_raw,
+    is_disabled,
+    provisioned=False,
+    provisioned_file=None,
+    updated_by=None,
+):
+    with Session(engine, expire_on_commit=False):
+        existing_workflow = get_workflow(tenant_id, id)
+        if not existing_workflow:
+            raise ValueError("Workflow not found")
+        return update_workflow_by_workflow(
+            existing_workflow,
+            name=name,
+            description=description,
+            interval=interval,
+            workflow_raw=workflow_raw,
+            is_disabled=is_disabled,
+            provisioned=provisioned,
+            provisioned_file=provisioned_file,
+            updated_by=updated_by,
+        )
+
+
+def update_workflow_by_workflow(
+    existing_workflow: Workflow,
+    name,
+    description,
+    interval,
+    workflow_raw,
+    is_disabled,
+    provisioned=False,
+    provisioned_file=None,
+    updated_by=None,
+):
+    # In case the workflow name changed to empty string, keep the old name
+    if name != "":
+        name = name
+    else:
+        name = existing_workflow.name
+    with Session(engine, expire_on_commit=False) as session:
+        # TODO: ensure that id is unique to workflow_id, like if rolled back and now updating from this version it should be bigger than latest version
+        next_revision = existing_workflow.revision + 1
+        updated_by_str = updated_by or existing_workflow.updated_by or "unknown"
+        # update version
+        version = WorkflowVersion(
+            workflow_id=existing_workflow.id,
+            revision=next_revision,
+            workflow_raw=workflow_raw,
+            updated_by=updated_by_str,
+            comment=f"Updated by {updated_by_str}",
+            # TODO: check if valid
+            is_valid=True,
+            is_current=True,
+        )
+        session.add(version)
+
+        existing_version = session.exec(
+            select(WorkflowVersion)
+            .where(WorkflowVersion.workflow_id == existing_workflow.id)
+            .where(WorkflowVersion.revision == existing_workflow.revision)
+        ).first()
+        if existing_version:
+            existing_version.is_current = False
+            session.add(existing_version)
+
+        # tb: no need to override the id field here because it has foreign key constraints.
```

Comment:
not sure what this comment was originally on, but it's not relevant here anymore